sbi/inference/snpe/snpe_base.py

```diff
@@ -232,58 +215,45 @@ def _get_log_prob_proposal_posterior(self, theta: Tensor, x: Tensor, masks: Tens
 
         Returns:
-            self._theta_bank: $\theta$ used for training
-            self._x_bank: $x$ used for training
-            self._prior_masks: Masks of 0/1 for each prior sample,
-                indicating whether prior sample will be used in current round
+            theta: Parameters used for training
+            x: Simulations used for training
+            prior_mask: Did the simulation come from a prior parameter sample?
         """
-        # Generate theta from prior in first round, and from most recent posterior
-        # estimate in subsequent rounds.
+
         if round_ == 0:
-            # New simulations are run only if pilot samples are not enough.
-            num_samples_remaining = num_samples - self._num_pilot_samples
-            if num_samples_remaining > 0:
-                theta, x = simulate_in_batches(
-                    simulator=self._simulator,
-                    parameter_sample_fn=lambda num_samples: self._prior.sample(
-                        (num_samples,)
-                    ),
-                    num_samples=num_samples_remaining,
-                    simulation_batch_size=self._simulation_batch_size,
-                )
-                theta = torch.cat((theta, self.pilot_theta[:num_samples]), dim=0,)
-                x = torch.cat((x, self.pilot_x[:num_samples]), dim=0,)
-            else:
-                theta = self.pilot_theta[:num_samples]
-                x = self.pilot_x[:num_samples]
+            # In round 0, mark simulations as made from prior-drawn parameters.
+            prior_mask = torch.ones((num_sims, 1), dtype=torch.bool)
+
+            num_missing_sims = max(num_sims - self._num_pilot_sims, 0)
+
+            # We take at most the total requested number from the pilot run.
+            theta = self.pilot_theta[:num_sims]
+            x = self.pilot_x[:num_sims]
+            if num_missing_sims > 0:
+                # If missing, we produce extra simulations as needed.
+                missing_theta = self._prior.sample((num_missing_sims,))
+                missing_x = self._batched_simulator(missing_theta, num_missing_sims)
+                theta = torch.cat((theta, missing_theta), dim=0)
+                x = torch.cat((x, missing_x), dim=0)
 
         else:
-            theta, x = simulate_in_batches(
-                simulator=self._simulator,
-                parameter_sample_fn=lambda num_samples: self._neural_posterior.sample(
-                    num_samples, x=self._x_o,
-                ),
-                num_samples=num_samples,
-                simulation_batch_size=self._simulation_batch_size,
-            )
+            # In subsequent rounds mark simulations as NOT from prior-drawn parameters.
+            prior_mask = torch.zeros((num_sims, 1), dtype=torch.bool)
+            # XXX Meke posterior.sample() accept tuples like prior.sample().
+            theta = self._neural_posterior.sample(num_sims, x=self._x_o)
```

Comment:
`x=self._x_o` not needed, should be default already

Reply:
you `self._neural_posterior` is instantiated with `self._x_o` as observation anyway, somewhere above? 

Maybe it is still good to leave this here for clarity? I dont know.

Reply:
Yes, I like it explicit. Leaving like it is, but can reopen discussion post-PR.